### PR TITLE
Terrain costs for 1849

### PR DESCRIPTION
**Pull request by tysen**
Transfer upgrade cost to newly placed tiles, charge discounted rates for
narrow and broad-placed-while-owning-P2

### CODE REVIEW

File: lib/engine/game/g_1849.rb [hunk 1]
```diff
@@ -49,6 +49,57 @@ def event_earthquake!
...
+                               Step::BuyTrain,
+                               [Step::BuyCompany, blocks: true],
+                             ], round_num: round_num)
+      end
+
+      def new_track(old_tile, new_tile)
+        # Assume path retention checked elsewhere
+        old_track = old_tile.paths.map(&:track)
+        added_track = new_tile.paths.map(&:track)
+        old_track.each { |t| added_track.slice!(added_track.index(t) || added_track.length) }
```
Comment (tobymao) [inline lib/engine/game/g_1849.rb:73]: always use size as opposed to length

File: lib/engine/game/g_1849.rb [hunk 2]
```diff
@@ -49,6 +49,57 @@ def event_earthquake!
...
+          added_track.include?(:broad) ? :broad : :narrow
+        end
+      end
+
+      def tile_cost(tile, _hex, entity, new_tile: nil)
+        return 0 if tile.upgrades.empty?
+
+        upgrade = tile.upgrades[0]
+        track = new_track(tile, new_tile)
+        case track
```
Comment (tobymao) [inline lib/engine/game/g_1849.rb:86]: simply to 

case new_track(...)

File: lib/engine/game/g_1849.rb [hunk 3]
```diff
@@ -49,6 +49,57 @@ def event_earthquake!
...
+        when :dual
+          upgrade.cost
+        when :narrow
+          @log << "#{entity.name} pays 1/4 cost for narrow gauge track"
+          upgrade.cost / 4
```
Comment (tobymao) [inline lib/engine/game/g_1849.rb:91]: is there rounding? care with division as js and ruby have different interpretations

File: lib/engine/part/path.rb [hunk 4]
```diff
@@ -61,7 +61,18 @@ def <=>(other)
 
       def <=(other)
         other_ends = other.ends
-        ends.all? { |t| other_ends.any? { |o| t <= o } }
+        ends.all? { |t| other_ends.any? { |o| t <= o } } && tracks_match(other.track)
+      end
+
+      def tracks_match(other)
```
Comment (tobymao) [inline lib/engine/part/path.rb:67]: use constants, i'm afraid this is a hot loop and will cerate too many objects

File: lib/engine/game/base.rb [hunk 5]
```diff
@@ -1034,7 +1034,7 @@ def place_home_token(corporation)
         city.place_token(corporation, token)
       end
 
-      def tile_cost(tile, hex, entity)
+      def tile_cost(tile, hex, entity, new_tile: nil) # rubocop:disable Lint/UnusedMethodArgument
```
Comment (tobymao) [inline lib/engine/game/base.rb:1037]: was new_tile really necessary? can't you just do hex.tile?

File: lib/engine/game/g_1849.rb (hunk 3 above)
Comment (tysen) [reply lib/engine/game/g_1849.rb:91]: No rounding - base costs are 160, 80, and 40

File: lib/engine/game/base.rb (hunk 5 above)
Comment (tysen) [reply lib/engine/game/base.rb:1037]: Good point. Didn't realize `hex.lay` had already been called by now.

**Review by tobymao — APPROVED** [review]

**Pull request merged**